File: csharp/meetup/Meetup.cs
using System;

public enum Schedule { First, Second, Third, Fourth, Last, Teenth };

public class Meetup
{
    private int year;
    private int month;

    public Meetup(int m, int y)
    {
        year = y;
        month = m;
    }

    public DateTime Day(DayOfWeek dayOfWeek, Schedule schedule)
    {
        var startOfMonth = new DateTime(year, month, 1);
        var day_offset = dayOfWeek - startOfMonth.DayOfWeek;
        if (day_offset < 0) {
            day_offset += 7;
        }
        switch (schedule) {
            case Schedule.First:
            case Schedule.Second:
            case Schedule.Third:
            case Schedule.Fourth:
                day_offset += (int)schedule * 7;
                break;
            case Schedule.Teenth:
                day_offset += 7;
                if (day_offset < 12)
                {
                    day_offset += 7;
                }
                break;
            case Schedule.Last:
                day_offset += 21;
                var daysInMonth = DateTime.DaysInMonth(year, month);
                if (daysInMonth - day_offset > 7)
                {
                    day_offset += 7;
                }
                break;
        }
        return startOfMonth.AddDays(day_offset);
    }
}

File: haskell/meetup/Meetup.hs
module Meetup(Weekday(..), Schedule(..), meetupDay) where

import Data.Time.Calendar(Day(..), fromGregorian, gregorianMonthLength, addDays)
import Data.Time.Calendar.WeekDate(toWeekDate)

data Weekday = Monday | Tuesday | Wednesday | Thursday | Friday | Saturday | Sunday deriving (Enum)

data Schedule = First | Second | Third | Fourth | Last | Teenth deriving (Enum)

firstDayOfMonth :: Integer -> Int -> Weekday
firstDayOfMonth y m = toEnum $ dayIdx - 1
  where (_,_,dayIdx) = toWeekDate $ fromGregorian y m 1

daysBetween :: Weekday -> Weekday -> Int
daysBetween from to = if d < 0 then d + 7 else d
  where d = fromEnum to - fromEnum from

getDaysToAdd :: Schedule -> Int -> Int -> Int
getDaysToAdd Last   d m = d + if d + 28 < m then 28 else 21
getDaysToAdd Teenth d _ = d + if d < 5 then 14 else 7
getDaysToAdd sched  d _ = d + 7 * fromEnum sched

meetupDay :: Schedule -> Weekday -> Integer -> Int -> Day
meetupDay schedule weekday year month = addDays daysToAdd firstOfMonth
  where firstOfMonth  = fromGregorian year month 1
        daysToAdd     = toInteger $ getDaysToAdd schedule daysToWeekday daysInMonth
        daysInMonth   = gregorianMonthLength year month
        daysToWeekday = daysBetween startDay weekday
        startDay      = firstDayOfMonth year month